### ISSUE TITLE
Sometimes doesn't load the image into the view on first call

### ISSUE THREAD

**Issue by eicm** [bumptech/glide#395]
In a PagerAdapter instantiateItem() I make the following call:

public Object instantiateItem(ViewGroup container, int position) {
...
ImageView imageView = new ImageView(getActivity());
...
Glide.with(getActivity()).load(path).crossFade().thumbnail(Glide.with(getActivity()).load(thumbPath).asBitmap().transcode(new BitmapToGlideDrawableTranscoder(getActivity()), GlideDrawable.class)).into(imageView);
...
}

Sometimes on the first call it doesn't set the image to the given imageView, but as I scroll far enough and come back it loads the image from cache for that given position which shows that the image has been cached on the first call.

What would be the reasons for not setting the image in the imageView in the first call.

**Comment by sjudd** [comment]
One possibility is that your view either doesn't undergo layout, or ends up with 0 width or 0 height. Glide tries to wait until you have a valid width and height before starting image loads so we know how to resize the image. 

Try using [``override()``][1] and see if that solves the problem. If it does, then it's almost certainly a view size issue.

[1]: http://bumptech.github.io/glide/javadocs/350/com/bumptech/glide/DrawableRequestBuilder.html#override(int,%20int)

**Comment by vaxus** [comment]
I've had this issue too a lot recently too. My imageviews have a predefined size (90dp x 90dp, not using wrap_content). 

If I back out of the listview and reload it or scroll up and down as mentioned by @eicm, the images sometimes load correctly. 

I've double checked the image links and there's no issue with them either.

Note: I'm using the viewholder pattern in my adapter to load images into my listview. 

Here's my adapter code:

ArrayList<Item> ItemArrayList;

    if (view == null) {

        LayoutInflater inflater = (LayoutInflater) viewGroup.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        row = inflater.inflate(R.layout.item_row, viewGroup, false);
        holder = new ItemHolder(row);
        row.setTag(holder);
    }

    else {
        row = view;
        holder = (ItemHolder) row.getTag();
    }

    Glide.with(viewGroup.getContext()).load(ItemArrayList.get(position).getItemImage()).transform(new GlideCircleTransform(c)).error(R.drawable.mm_avatar).into(holder.itemImage);

**Comment by vaxus** [comment]
Quick update. After enabling adb debugging as mentioned in the docs, I am getting the following error message on failed image loads:

    java.lang.IllegalStateException: Can't call reconfigure() on a recycled bitmap

Any ideas how to go about solving this?

**Comment by sjudd** [comment]
@vaxus that sounds like you're returning a bitmap to the bitmap pool or recycling it twice. Do you still see the error if you remove your Transformation from your load call?

**Comment by eicm** [comment]
@sam setting a fixed size seems to solve the problem, though I need to
verify it more.

- on another topic loading MediaStore thumbnails using thumbnail's path and
load() is returning considerably faster than using loadFromMediaStore() and
using original_id of the image.

**Comment by vaxus** [comment]
@sjudd I removed the transformation from the load call and it still gives me:
    java.lang.IllegalStateException: Can't call reconfigure() on a recycled bitmap

**Comment by sjudd** [comment]
Interesting, can you provide a stack trace? I don't see anything else obviously wrong here. I may need a sample app to look into this further...

Interesting that just using the thumbnail path is faster than loadFromMediaStore. Do you have media store thumbs for those files? If not, loadFromMediaStore will fall back to the file path. It's possible that when you don't have thumbs the queries for the thumb file path take long enough to significantly slow down the load.

**Comment by eicm** [comment]
I noticed my target size has bigger width and length compared to the
MINY_THUMBS width and height and that's why in loadFromMediaStore it
actually loaded down sampled version of the original images, which as you
mentioned earlier I can use override() to set the target manually.

**Comment by eicm** [comment]
The other thing that I noticed is that the orientation is not taken into
the account for loading MediaStore thumbnails when loaded using
loadFromMediaStore while it has been taken into the account for to rotate
the images regarding their orientation when loading original images.

**Comment by sjudd** [comment]
@eicm It's possible different devices behavior differently w/r/t thumbnail orientation. The code path for media store thumbs and full size images is actually the same. The only difference is that for media store thumbs, we load data from the thumb file path rather than the original image file path.

As a result, if the thumb contains exif data or is encoded with the correct orientation, we should provide a correctly rotated image.

What device/version of Android do you see incorrectly rotated thumbs on?

**Comment by eicm** [comment]
I'm experiencing it on all devices I have: nexus 10 4.4, nexus 9 5.0.1,
lenova 4.2 and kindle fire OS. The rotated images have orientation set in
exif and also MediaStore orientation column. What I'm doing now is to
rotate the thumbs manually in the SimpleImage call back. But it's loading
the full size images correctly. Could it be because I'm using override
(mini_width, mini_height) in loadFromMediaStore() to force it loading the
thumbnails? (My target size is bigger than android's mini thumbs, and I'm
first loading mini thumbs and then images in my target)

**Comment by sjudd** [comment]
I looked into this, it turns out MediaStore.Images.Thumbnails neither produces correctly oriented images nor includes the original exif orientation tag in the thumbnail.

For now you can work around this by using ``load(Uri)`` or ``fromUri()`` instead of ``loadFromMediaStore()``.

I'll try and think of sane way to pass through the original orientation...

**Comment by vaxus** [comment]
I'm unable to provide a sample app unfortunately. I tried to use Picasso just to see if I would experience similar behaviour and it seemed to work ok. I'm not getting the errors during loads. Very strange :/

**Comment by sjudd** [comment]
@vaxus Can you file a new issue for the IllegalStateException?

**Comment by sjudd** [comment]
I'm going to close this issue since I think we addressed the original problem (images aren't loading). I filed a separate issue for the media store thumbnail rotation issue (#400) and we should file another for the IllegalStateException.

**Comment by TheAndroidMaster** [comment]
I'm getting this issue a lot now. I'm not sure if it is because of the same reason, though. I tried to get a lot of images at once and create views for them in a staggered list view of sorts (two linearlayouts). I have a void (getDominantColor) that finds the average color of the image and sets it as the background when the image is loaded. It was when this code was added that the error started happening. Here is my code:

        count=0;
        isOdd=true;

        while(count<name.length){
            final CardView cardView = new CardView(Flat.this);
            cardView.setClickable(true);
            cardView.setUseCompatPadding(true);
            cardView.setId(count);
            cardView.setLayoutParams(new CardView.LayoutParams(CardView.LayoutParams.MATCH_PARENT, CardView.LayoutParams.WRAP_CONTENT, 1));
            cardView.setForeground(getSelectedItemDrawable());
            String mName = name[count];
            mName = mName.replace("*", "");
            LinearLayout vertical = new LinearLayout(Flat.this);
            vertical.setOrientation(LinearLayout.VERTICAL);
            final ImageView image = new ImageView(Flat.this);
            image.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 100));
            image.setBackgroundColor(getResources().getColor(R.color.oranged));
            LinearLayout horizontal = new LinearLayout(Flat.this);
            horizontal.setOrientation(LinearLayout.HORIZONTAL);
            horizontal.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
            horizontal.setId(count);
            final TextView title = new TextView(Flat.this);
            title.setBackgroundColor(getResources().getColor(R.color.orange));
            title.setTextColor(getResources().getColor(R.color.white));
            title.setPadding(50, 40, 50, 40);
            title.setText(mName);
            Picasso.with(this)
                    .load(url[count])
                    .resize(100, 100)
                    .centerInside()
                    .into(new Target() {
                        @Override
                        public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                            image.setBackgroundColor(getDominantColor(bitmap));
                            title.setBackgroundColor(getDominantColor(bitmap));
                            image.setImageBitmap(bitmap);
                        }

                        @Override
                        public void onBitmapFailed(Drawable errorDrawable) {

                        }

                        @Override
                        public void onPrepareLoad(Drawable placeHolderDrawable) {

                        }
                    });
            cardView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(Flat.this, WallpaperView.class);
                    intent.putExtra("num", v.getId());
                    intent.putExtra("array", whichname);
                    intent.putExtra("arrays", whichurl);
                    intent.putExtra("up", "Flat");
                    Flat.this.startActivity(intent);
                }
            });
            if(isOdd){
                root.addView(cardView);
                cardView.addView(vertical);
                vertical.addView(image);
                vertical.addView(title);
                isOdd=false;
            }else{
                rootView.addView(cardView);
                cardView.addView(vertical);
                vertical.addView(image);
                vertical.addView(title);
                isOdd=true;
            }
            count=count+1;
        }

**Comment by TWiStErRob** [comment]
@TheAndroidMaster you should open a new issue if you're really having this issue based on our [issue template](https://github.com/bumptech/glide/blob/master/CONTRIBUTING.md#filing-issues) (there were many fixes since this issue has been closed), **but before you do that please try to use Glide ;)**

Unrelated: you should look into `LayoutInflater.from(Context)`, it's easier to maintain xml layouts than code-generating the same (mostly because you have preview). You can still `setId` dynamically on the inflated view if need be.

**Comment by aLoyalStudent** [comment]
I have the same issues.When I load image in the listView header,the first time ,Glide not set image to imageView.when I refresh the ui,it is normal.why ? thanks.

**Comment by TheAndroidMaster** [comment]
Ohgod I forgot I posted here please don't look at my horrible code
_I definitely wasn't loading views into two LinearLayouts instead of using a RecyclerView_ xD

**And I wasn't even using Glide when I posted this what the heck was I thinking!?**

@aLoyalStudent An example of how you're loading the image might be useful ;)

**Comment by TWiStErRob** [comment]
@aLoyalStudent please open a new issue and fill in the details!
